url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/environment/geometry/Cone.java

```diff
@@ -0,0 +1,344 @@
+        //- Calcula el termino B --------------------------------------------
+        B = 2 *
+            ((r.direction.x * r.origin.x) +
+             (r.direction.y * r.origin.y) -
+             (r.direction.z * r.origin.z * VSDK.square(inR)) / 
+             VSDK.square(inH)
+             );
+
+        //- Calcula el termino C --------------------------------------------
+        C = VSDK.square(r.origin.x) +
+            VSDK.square(r.origin.y) -
+            VSDK.square(r.origin.z * inR / inH);
+
+        //- Calcula el discriminant. Si el discriminant no es positivo el -
+        //- rayo no intersecta la esfera. retorna t = 0                      
+        discriminant = VSDK.square(B) - 4*A*C;
+        if ( discriminant <= VSDK.EPSILON ) return false;
+
+        //- Resuelve la ecuacion cuadratica para las raices de la ecuacion. -
+        //- (-B +/- sqrt(B^2 - 4*A*C)) / 2A.                                -
+        discriminant = Math.sqrt(discriminant);
+        t0 = (-B-discriminant) / (2 * A);
+
+        //- Si t0 es > 0 listo. Si no debemos calcular la otra raiz t1. -----
+        if ( t0 > VSDK.EPSILON ) {
+            // OJO: Aqui va el calculo del punto y la normal!
+            outInfo.p = r.origin.add(r.direction.multiply(t0));
+            if ( outInfo.p.z > 0 || outInfo.p.z < -inH ) {
+                return false;
+            }
+
+            // Se calcula la normal como el gradiente de la formula del cono,
+            // notese que aqui se obtiene un vector escalado en 1/2 respecto al
+            // gradiente (para ahorrar multiplicaciones)
+            outInfo.n.x = outInfo.p.x;
+            outInfo.n.y = outInfo.p.y;
+            outInfo.n.z = -outInfo.p.z * VSDK.square(inR/inH);
+            outInfo.n.normalize();
+    
+            outInfo.p.z += inH;
+            inOutRay.t = t0;
+
+            return true;
+        }
+
+        return false;
+
+    }
+
+    private boolean
+    doIntersectionTap(Ray inOutRay, double inR, double inH,
+                      GeometryIntersectionInformation outInfo) {
+        double t;
+        Vector3D p;
+        Vector3D proy;
+        Vector3D o = new Vector3D(0, 0, 0);
+
+        if ( Math.abs(inOutRay.direction.z) > VSDK.EPSILON ) {
+            t = (inH - inOutRay.origin.z) / inOutRay.direction.z;
+            if ( t > VSDK.EPSILON ) {
+                p = inOutRay.origin.add(inOutRay.direction.multiply(t));
+                proy = new Vector3D(p.x, p.y, 0);
+                if ( VSDK.vectorDistance(proy, o) < inR ) {
+                    inOutRay.t = t;
+                    outInfo.n.x = 0;
+                    outInfo.n.y = 0;
+                    outInfo.n.z = 1;
+                    outInfo.p = p;
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     Check the general interface contract in superclass method
+     Geometry.doIntersection.
+    */
+    public boolean
+    doIntersection(Ray inOutRay) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 28. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 241-261 into a separate method.